Create graceful-delegate-shutdown-process.md (#575)
harness/developer-hub: docs/platform/2_Delegates/get-started-with-delegates/graceful-delegate-shutdown-process.md

---
title: Graceful delegate shutdown

> description: This document explains how delegate shutdown works based on delegate type and environment.

Reviewer: ```suggestion
description: Learn how delegate shutdown works based on delegate type and environment.
```

Reviewer: just suggesting a shorter description since the cards in our doc site are so small.

Author: I think if I change it I'll stick in the term graceful since this document does not explain much more than that. But will wait to see what an SME says about the below changes before I controvert your approval.

# sidebar_position: 2
---

> Harness Delegate is implemented to shut down gracefully. The process of graceful shutdown is applied as follows:
> 
> - Delegate receives an instruction to quit.
> - A grace period begins during which the delegate:
>   + Stops accepting new tasks.
>   + Works to complete running tasks.
> - The grace period ends.
> - Delegates that have not quit are force-terminated.
> - Incomplete tasks are discarded.

Reviewer: ```suggestion
Harness Delegate shuts down gracefully. The shutdown process follows these steps:

1. Delegate receives an instruction to quit.
2. A grace period begins during which the delegate:
  + Stops accepting new tasks.
  + Works to complete running tasks.
3. The grace period ends.
4. Delegates that have not quit are force-terminated.
5. Incomplete tasks are discarded.
```

Reviewer: seems like this is a sequence so an ordered list is more useful.

Author: Always bulleted list when you're just describing a process. At least that's the way I've been taught. Numbers are for procedures with discrete steps.

> The length of the grace period is configurable and depends on whether the delegate is immutable or legacy. 

Reviewer 2: remove the context to legacy, just say delegate with the immutable image

Author: @pratmit This was changed to "The length of the grace period is configurable."

> The length of the grace period is configurable and depends on whether the delegate is immutable or legacy. 

Author: ```suggestion
The length of the grace period is configurable. 
```

| **Delegate type** | **Grace period** | **Default interval** |
| :-- | :--: | :--: |
| Immutable | Yes | 10 minutes |
| Legacy | No | 30 seconds |

The event that initiates the graceful shutdown depends on delegate type.

| **Delegate environment** | **Trigger** |

> | Kubernetes | Pod termination, eviction, or user-initiated scaling |

Reviewer: Sure about this?

The K8s Delegate is a deployment object. 

When a Kubernetes pod is terminated in a Deployment object, the Deployment's replica count is decremented and a new pod is created to replace the terminated pod. The new pod will have the same configuration as the previous pod and continue to run the same containers and associated processes. This is because Deployments are designed to ensure that the desired number of replicas are running at all times, and to provide automatic self-healing of pods in case of failures or terminations.

I would image you would need to terminate the Deployment to initiate shutdown. Like we cover here:

https://developer.harness.io/docs/platform/delegates/manage-delegates/delete-a-delegate/

Author: I'm familiar with how replica sets work but from what I see here, I think pod termination could still trigger a graceful delegate shutdown process. Raghu or another SME from the team will review this. Thank you.

| Docker | `docker kill` command |

> | Shell | `./stop.sh` instruction |

Reviewer: we have a shell script delegate now? We did in FG, but I didn't know it was in NG.

Author: Yeah, I can't really say. This was given to me and, as you know, I'm still learning. But I'm having a technical reviewer look at it and the PM will take a second look too. I think, actually, that this information was from the issue and that that ticket was opened by a technical lead.

---
title: Graceful delegate shutdown
description: Read about the process of graceful delegate shutdown.
# sidebar_position: 2
---

Harness Delegate is designed to shut down gracefully. The process of graceful shutdown proceeds as follows:

- The delegate receives an instruction to quit.
- A grace period begins during which the delegate:
  + Stops accepting new tasks.
  + Works to complete running tasks.
- The grace period ends.
- Delegates that have not quit are force-terminated.
- Incomplete tasks are discarded.

The length of the grace period is configurable. 

| **Delegate type** | **Grace period** | **Default interval** |
| :-- | :--: | :--: |

> | Immutable | Yes | 10 minutes |

Reviewer 2: Immutable Image

> | Immutable | Yes | 10 minutes |

Author: ```suggestion
| Immutable image | Yes | 10 minutes |
```

> | Legacy | No | 30 seconds |

Reviewer 2: Legacy Image

> | Legacy | No | 30 seconds |

Author: ```suggestion
| Legacy image | No | 30 seconds |
```
